extensions/kafka/src/main/java/com/hazelcast/jet/kafka/KafkaDataConnection.java

```diff
@@ -138,11 +155,53 @@ public <K, V> KafkaProducer<K, V> getProducer(@Nullable String transactionalId)
         }
     }
 
+    /**
+     * Returns an instance of {@link KafkaProducer} based on the DataConnection
+     * configuration and given properties.
+     * <p>
+     * The caller is responsible for closing the producer instance.
+     * For non-shared producers the producer will be closed immediately
+     * upon calling close.
+     * For shared producers the producer will be closed when all users
+     * close the returned instance and this DataConnection is released by
+     * calling {@link #release()}
+     * <p>
+     *
+     * @param transactionalId transaction id to pass as 'transactional.id'
+     *                        property to a new KafkaProducer instance,
+     *                        must be null for shared producer
+     * @param properties      properties. E.g, SQL mappings provide separate
+     *                        options, and they should be merged with
+     *                        data connection's properties. These properties
+     *                        has higher priority than data connection properties.
+     */
+    @Nonnull
+    public <K, V> KafkaProducer<K, V> getProducer(
+            @Nullable String transactionalId,
+            @Nonnull Properties properties) {
+        if (getConfig().isShared()) {
+            throw new HazelcastException("Shared Kafka producer can be created only with data connection options");
+        } else {
+            Properties props = mergeProps(properties);
+            if (transactionalId != null) {
+                props.put("transactional.id", transactionalId);
+            }
+            return new KafkaProducer<>(props);
+        }
+    }
+
+    Properties mergeProps(Properties properties) {
+        Properties props = new Properties();
+        props.putAll(getConfig().getProperties());
+        props.putAll(properties);
+        return props;
+    }
+
     @Override
     public void destroy() {
-        if (producer != null) {
-            producer.doClose();
-            producer = null;
+        if ((producerSupplier != null) && (producerSupplier.remembered() != null)) {
+            producerSupplier.remembered().doClose();
```

Comment:
this is not a standard usage for `ConcurrentMemoizingSupplier`. Is it thread safe?

Reply:
we need a `forgetAndDoWithForgottenObject(Consumer<T>)` method

Reply:
is `destroy` method actually required to be thread-safe? I read the code and docs and seems that it is invoked only once for given object

Reply:
it seems it can be invoked multiple times when the node is shutting down